core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala

```diff
@@ -199,14 +201,25 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, clock: Clock)
       if (now - lastSeenMs > executorTimeoutMs) {
         logWarning(s"Removing executor $executorId with no recent heartbeats: " +
           s"${now - lastSeenMs} ms exceeds timeout $executorTimeoutMs ms")
-        scheduler.executorLost(executorId, SlaveLost("Executor heartbeat " +
```

Comment:
does it hurt to call `scheduler.executorLost` twice? Since we assume killing executors may take a while, looks reasonable to call `scheduler.executorLost` and re-submit tasks eariler.

Reply:
No serious hurt, but gives waring/error says that the executor has already been removed, etc.

Reply:
> Since we assume killing executors may take a while, looks reasonable to call scheduler.executorLost and re-submit tasks earlier.

I noticed a wired problem of this. That is when we call `scheduler.executorLost` in `HeartbeatReceiver`, `scheduler` knows which executor lost, but `CoarseGrainedSchedulerBackend` doesn't. So, even if we want to re-submit tasks earlier, tasks can still be launched on that lost executor.

But for other two places where we call `scheduler.executorLost`, it guarantees that `CoarseGrainedSchedulerBackend` has realized about the lost/killed executor.

So, I think that `HeartbeatReceiver` may has used it incorrectly, here.

Reply:
makes sense!

Reply:
The problem is we should avoid call TaskScheduler directly here, instead we shall shadow it behind the SchedulerBackend, to make resource information tracking  easier.